### PR TITLE
Cancelling the running task (executing of a scaffolder template)

### PR DESCRIPTION
**Pull request by acierto**
## Hey, I just made a Pull Request!

As a "Scaffolder template" user I would like to have a possibility to abort the current task for various reasons:

I realised that ran it with wrong input
I found out that something is going wrong
Task is hanging and I would like to abort it immediately (not waiting for a scheduler to vacuum it).

Screenshot 2023-02-23 at 08 54 43
Screenshot 2023-02-23 at 08 55 07
Screenshot 2023-02-23 at 08 55 22
Screenshot 2023-02-25 at 22 31 41

Test case is done based on this template: https://github.com/acierto/software-templates/blob/main/scaffolder-templates/docs-long-running-template/template.yaml by running on 2 instances of backstage. In instance 1 the task is running, and instance 2 is cancelling it. 

#### :heavy_check_mark: Checklist

- [x] A changeset describing the change and affected packages. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#creating-changesets))
- [x] Added or updated documentation
- [x] Tests for new functionality and regression tests for bug fixes
- [x] Screenshots attached (for UI changes)
- [x] All your commits have a `Signed-off-by` line in the message. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#developer-certificate-of-origin))

### CODE REVIEW

**Comment by acierto** [comment]
> ## Uffizzi Preview Environment `deployment-9846`
> ☁️ https://app.uffizzi.com/github.com/backstage/backstage/pull/15461
> 
> 📄 [View Application Logs etc.](https://app.uffizzi.com/projects/4436/deployments/9846/containers)
> 
> What is Uffizzi? [Learn more](https://github.com/UffizziCloud/uffizzi)

Some configuration is missing with Uffizzi, it doesn't read pick up templates 😢

**Comment by benjdlambert** [comment]
> > ## Uffizzi Preview Environment `deployment-9846`
> > ☁️ https://app.uffizzi.com/github.com/backstage/backstage/pull/15461
> > 📄 [View Application Logs etc.](https://app.uffizzi.com/projects/4436/deployments/9846/containers)
> > What is Uffizzi? [Learn more](https://github.com/UffizziCloud/uffizzi)
> 
> Some configuration is missing with Uffizzi, it doesn't pick up templates 😢

Yeah I need to add the templates to the config I think :(

**Comment by acierto** [comment]
@Rugvip 

The link to the original PR: https://github.com/backstage/backstage/pull/14388 
It was auto-closed during my holidays. 
The comments were incorporated. 

At the moment of clicking on "Abort" button, the task step is marked for abortion and the task runner will abort it eventually. 
Meanwhile user will see on the "Aborting", as an indication that the abortion is in action.

**Comment by acierto** [comment]
Is there anything what can I do to simplify the review of the PR? 
Some flow drawing or automation test

File: plugins/scaffolder-backend/src/scaffolder/tasks/StorageTaskBroker.ts
Comment (freben) [inline plugins/scaffolder-backend/src/scaffolder/tasks/StorageTaskBroker.ts:69]: No need for this; you already have `ac.signal.aborted` to use at any point. So I'd say, just return the raw AbortController from this method instead of wrapping it up, and skip the isAborted member, probably remove the `aborted` method below too then right since it just duplicates information?

File: plugins/scaffolder/src/components/hooks/useEventStream.ts
Comment (freben) [inline plugins/scaffolder/src/components/hooks/useEventStream.ts:105]: makes me wonder what the distinction is between cancelled and aborted 🤔

File: plugins/scaffolder-backend/src/scaffolder/tasks/types.ts
Comment (freben) [inline plugins/scaffolder-backend/src/scaffolder/tasks/types.ts:106]: Let's try to get rid of this type

File: plugins/scaffolder-backend/src/scaffolder/tasks/TaskWorker.ts
Comment (freben) [inline plugins/scaffolder-backend/src/scaffolder/tasks/TaskWorker.ts:116]: so could it happen that this subscription lingers forever (potential memory leak right?) if an error occurs that's not followed by a regular message?

File: plugins/scaffolder-backend/src/scaffolder/tasks/TaskWorker.ts
Comment (freben) [inline plugins/scaffolder-backend/src/scaffolder/tasks/TaskWorker.ts:141]: Not fond of the cast here - that might just shadow over actual problems down the line, especially if the manager is replaceable. Should be possible to find the right abort controller anyway.

Honestly, why do we even listen to a broker event stream? The broker is the one that's responsible for responding to the claim(). Why doesn't the broker return a context that has an `abortSignal: AbortSignal` in it? The broker is the one that knows when it's time to abort, isn't it, based on its internal functionality? So it can hold the corresponding `AbortController` internally, and trigger it at the right time which has the side effect that this worker also gets informed about it through the signal

File: plugins/scaffolder-backend/src/scaffolder/tasks/TaskWorker.test.ts
Comment (freben) [inline plugins/scaffolder-backend/src/scaffolder/tasks/TaskWorker.test.ts:327]: Another side note here I think, is that getWorkspaceName _happens_ to return the task ID. It's not documented or settled behavior. Using it to deduce the task ID is risky and subject to future breakage.

**Review by freben — COMMENTED** [review]
Alright some open questions, sorry if I sound confused - I did get a bit surprised by the shape of the solution, but totally open to that it might just be me :) @benjdlambert ping here too

File: plugins/scaffolder-node/src/actions/types.ts
Comment (acierto) [reply plugins/scaffolder-node/src/actions/types.ts:67]: Yeah, if we will make not optional now, we have to implement it in all actions plus we will force all users with custom actions to implement it 😄

Comment (freben) [reply plugins/scaffolder-node/src/actions/types.ts:67]: Do we though? You don't HAVE to listen to the signal right

Comment (acierto) [reply plugins/scaffolder-node/src/actions/types.ts:67]: It's not about listening the event (that's why it's optional), but specifying a signal as a parameter in each action. 

For example, one of actions is 'publish:gitlab:merge-request', with non-optional signal we have to define it in context

Screenshot 2023-02-16 at 19 11 47

And like that in each action

File: plugins/scaffolder-backend/src/scaffolder/tasks/TaskWorker.ts
Comment (acierto) [reply plugins/scaffolder-backend/src/scaffolder/tasks/TaskWorker.ts:141]: Ok, I will try to think about other way how to access abort controller if that's possible.

Regarding listening an event steam. It's related to a running Backstage in a cluster, where the task is running in instance A, but user can abort it from instance B.

File: plugins/scaffolder/src/components/hooks/useEventStream.ts
Comment (acierto) [reply plugins/scaffolder/src/components/hooks/useEventStream.ts:105]: I can use 'cancelled', but then we can also rename the action to "Cancel", not "Abort" :) to avoid confusion later.
Seems like 'cancelled' is not used so far

File: plugins/scaffolder-backend/src/scaffolder/tasks/types.ts
Comment (acierto) [reply plugins/scaffolder-backend/src/scaffolder/tasks/types.ts:208]: Body is here only for making the more informative message.

```
await this.storage.abortTask({
      taskId,
      body: {
        message: `Step ${currentStepId} has been aborted.`,
        ...{
          stepId: currentStepId,
          status: 'aborted',
        },
      },
    });
```

File: plugins/scaffolder-backend/src/scaffolder/tasks/TaskWorker.ts
Comment (acierto) [reply plugins/scaffolder-backend/src/scaffolder/tasks/TaskWorker.ts]: abort-task drawio

Comment (freben) [reply plugins/scaffolder-backend/src/scaffolder/tasks/TaskWorker.ts]: Ah, right. But no matter how the information travels, the task broker in Backend1 still is the one who would know how to abort the task it holds. So it should be kept as a broker-internal concern, and it only injecting a signal (not controller) into its in-memory task

Comment (acierto) [reply plugins/scaffolder-backend/src/scaffolder/tasks/TaskWorker.ts]: Yeah, I agree. I already have a working solution for that. 
Have to fix a test yet.

File: plugins/scaffolder/src/components/hooks/useEventStream.ts
Comment (freben) [reply plugins/scaffolder/src/components/hooks/useEventStream.ts:105]: Cancel is a pretty good word. Abort has some negative connotations and used less in regular speech.

Comment (acierto) [reply plugins/scaffolder/src/components/hooks/useEventStream.ts:105]: Okie dok. I'll change all to "Cancel".
I can't run backstage anymore after merged master to my branch.

Will wait when it be fixed first.
Seems like something is wrong with transpiler configuration for this plugin.

Screenshot 2023-02-22 at 21 52 22

File: app-config.yaml
Comment (freben) [inline app-config.yaml:277]: Let's not add this to the actual app config for everyone

File: plugins/scaffolder-backend/api-report.md
Comment (freben) [inline plugins/scaffolder-backend/api-report.md:742]: you can just do `.abortSignal.aborted` instead right, so we keep the information in one place

File: plugins/scaffolder-backend/api-report.md
Comment (freben) [inline plugins/scaffolder-backend/api-report.md:740]: If the wording everywhere else is "cancel", is it worth calling this `cancelSignal`? I'm not sure which is more confusing :)

File: plugins/scaffolder-backend/src/scaffolder/dryrun/createDryRunner.ts
Comment (freben) [inline plugins/scaffolder-backend/src/scaffolder/dryrun/createDryRunner.ts:118]: this comment used to apply to `done`

File: plugins/scaffolder/src/components/TaskPage/TaskPage.tsx
Comment (freben) [inline plugins/scaffolder/src/components/TaskPage/TaskPage.tsx:260]: Would you mind making this just `props: TaskPageProps` and unpacking inside the function body instead? Makes the API report much cleaner.

**Review by freben — COMMENTED** [review]
Alright, getting close!

File: plugins/scaffolder-backend/src/scaffolder/actions/builtin/debug/wait.ts
Comment (acierto) [reply plugins/scaffolder-backend/src/scaffolder/actions/builtin/debug/wait.ts:45]: Oeps

File: app-config.yaml
Comment (acierto) [reply app-config.yaml:277]: Oh sorry, that was unintended commit

**Comment by freben** [comment]
I see there are still conflicts. Would it be worth actually squashing all of this into a single commit, and rebasing properly on top of master instead? Backward merges like this can be notoriously tricky. The git tree becomes increasingly tangled and the conflicts ever harder to fix

**Comment by acierto** [comment]
> I see there are still conflicts. Would it be worth actually squashing all of this into a single commit, and rebasing properly on top of master instead? Backward merges like this can be notoriously tricky. The git tree becomes increasingly tangled and the conflicts ever harder to fix

Yeah, for the last > 1 week some commits to master created conflicts 😅 
Squashed it into one commit as you suggested.

File: plugins/scaffolder-backend/src/scaffolder/tasks/NunjucksWorkflowRunner.ts
Comment (freben) [inline plugins/scaffolder-backend/src/scaffolder/tasks/NunjucksWorkflowRunner.ts:189]: No changes in here right? Made it a bit harder to review; had to carefully manually diff the two :)

File: plugins/scaffolder/src/components/TaskPage/TaskPage.tsx
Comment (freben) [inline plugins/scaffolder/src/components/TaskPage/TaskPage.tsx:244]: These might have been unnecessary at this point. If there's no direct use case for them immediately, it may be better to wait for proper i18n support to replace strings like these all over the place in my humble opinion. I see that there was already a text for loading, so I presume you were being diligent and following the pattern, so I'll ask @benjdlambert

**Review by freben — COMMENTED** [review]
Alright I'm happy after these small remainders! Handing over to @benjdlambert for final approval.

File: plugins/scaffolder-backend/src/scaffolder/tasks/NunjucksWorkflowRunner.ts
Comment (acierto) [reply plugins/scaffolder-backend/src/scaffolder/tasks/NunjucksWorkflowRunner.ts:189]: No changes indeed, I extracted the step execution for easier readability.

File: docs/features/software-templates/writing-custom-actions.md
Comment (benjdlambert) [inline docs/features/software-templates/writing-custom-actions.md:72]: Think this is a bad merge conflict. This stuff still wants to be here.

File: docs/features/software-templates/writing-custom-actions.md
Comment (benjdlambert) [inline docs/features/software-templates/writing-custom-actions.md:74]: Not sure we should be documenting `node:`yet. We don't anywhere else and it's just confusing I think at this point.

File: docs/features/software-templates/writing-custom-actions.md
Comment (benjdlambert) [inline docs/features/software-templates/writing-custom-actions.md:73]:
```suggestion
import { createTemplateAction } from '@backstage/plugin-scaffolder-node';
```

File: plugins/scaffolder-backend/src/scaffolder/tasks/TaskWorker.test.ts
Comment (benjdlambert) [inline plugins/scaffolder-backend/src/scaffolder/tasks/TaskWorker.test.ts:318]: Hmm do we need this?

File: plugins/scaffolder/src/next/OngoingTask/OngoingTask.tsx
Comment (benjdlambert) [inline plugins/scaffolder/src/next/OngoingTask/OngoingTask.tsx:106]: Would be good to write some tests for this too.

**Review by benjdlambert — CHANGES_REQUESTED** [review]
Hey :wave: sorry first time having a glance at this after all the previous changes. Looking pretty good at the minute, got some higher level questions about the api changes :)

File: docs/features/software-templates/writing-custom-actions.md
Comment (acierto) [reply docs/features/software-templates/writing-custom-actions.md:74]: hm, ok, I thought to give some example  how to do such kind of aborting stuff, changed it to `node:fs` as `fs-extra` doesn't support abortable

File: plugins/scaffolder-node/api-report.md
Comment (acierto) [reply plugins/scaffolder-node/api-report.md:33]: I just changed it from `signal?` to `signal` on request from @freben 🤣

File: docs/features/software-templates/writing-custom-actions.md
Comment (benjdlambert) [reply docs/features/software-templates/writing-custom-actions.md:74]: you can just do `fs` instead of `node:fs` :)

File: plugins/scaffolder-node/api-report.md
Comment (benjdlambert) [reply plugins/scaffolder-node/api-report.md:33]: Ah I thought i'd seen that before, let me go read a little bit.

Comment (freben) [reply plugins/scaffolder-node/api-report.md:33]: I'm perfectly fine to let you decide though :)

Comment (freben) [reply plugins/scaffolder-node/api-report.md:33]: I'm perfectly fine to let you decide though :)

Comment (acierto) [reply plugins/scaffolder-node/api-report.md:33]: Initially i implemented it as an optional due to the same reason, to not introduce breaking changes.
But after @freben  pointed it out to change to non-optional, I reviewed it again, and indeed the context is defined in `NunjucksWorkflowRunner` and it won't do any breaking changes, only for unit tests.

File: plugins/scaffolder/src/next/OngoingTask/OngoingTask.tsx
Comment (acierto) [reply plugins/scaffolder/src/next/OngoingTask/OngoingTask.tsx:64]: I can move it, why i place it here is to make the calculation of "cancelability" at one place.

```
const cancelEnabled = !(
    taskStream.cancelled ||
    clickedToCancel ||
    taskStream.completed
  );
```

If we will move this line into `ContextMenu`, it will make to read the code a bit harder, is it option disabled now or not...

File: plugins/scaffolder-backend/src/scaffolder/tasks/TaskWorker.test.ts
Comment (acierto) [reply plugins/scaffolder-backend/src/scaffolder/tasks/TaskWorker.test.ts:318]: Yeah, as we don't need to define for the test all method definitions. It would be harder to read it then.

It's interesting, that if only claim method is defined, type system is happy

Screenshot 2023-03-08 at 19 54 33

but if to add also $event, now type checks wants all methods to be defined 🤯 

Screenshot 2023-03-08 at 19 54 42

**Comment by acierto** [comment]
@benjdlambert solved most of comments, for some left questions/remarks

File: plugins/scaffolder/src/next/OngoingTask/OngoingTask.tsx
Comment (acierto) [reply plugins/scaffolder/src/next/OngoingTask/OngoingTask.tsx:64]: I tried to move it, and then we need pass `taskStream` inside ContextMenu

Screenshot 2023-03-08 at 21 24 59

I can move it inside, but it will add unnecessary logic to ContextMenu. 
Up to you.

File: plugins/scaffolder-node/api-report.md
Comment (benjdlambert) [reply plugins/scaffolder-node/api-report.md:33]: I guess i'm thinking that if we've had to update tests to add this to the `ActionContext` it's breaking change as others will copy the implementation to test their own actions. I know I've written actions doing this, so it's unfortunate that it's required. Apologies for the confusion, but I think that whilst these are exported we should hold ourself accountable for changes :)

File: plugins/scaffolder/src/next/OngoingTask/OngoingTask.tsx
Comment (benjdlambert) [reply plugins/scaffolder/src/next/OngoingTask/OngoingTask.tsx:64]: I think you can move the `useCallback` and and the state inside the button, and also be able to drive the state from a prop too, something like `disabled` or something?

**Comment by benjdlambert** [comment]
Whilst this is getting close, think that we should hold off on merging this until after the mainline release which goes out next week. Feeling its a little risky to ship it and not test it in a next release beforehand.

**Comment by freben** [comment]
@acierto I see you've been affected by our switch to the docusaurus 2 microsite. Just make sure that those doc changes go in the right place in the `microsite` directory (the `microsite-next` is gone) and you should be fine

File: microsite/pages/en/plugins.js
Comment (freben) [inline microsite/pages/en/plugins.js:1]: What's up with this new empty file?

**Review by freben — APPROVED** [review]
Alright! I remove the extra `microsite/pages/en/plugins.js` that was added by mistake, now just waiting for it to go green!

**Review by benjdlambert — APPROVED** [review]

**Pull request merged**
